feat: ArtifactRepositoryRef ConfigMap is now taken from the workflow namespace

Ark-kun opened the pull request:
Previously, it used the controller's namespace.

gaoning777 commented:
/lgtm

alexec commented:
@jessesuen this PR changes to use the workflow's namespace for the artifact repo. I'm pretty sure this is a breaking change. Can I get your input please?

Ark-kun commented:
This test failure seems to be a fluke. @sarabala1979 is it possible to rerun them?

> @jessesuen this PR changes to use the workflow's namespace for the artifact repo. I'm pretty sure this is a breaking change. Can I get your input please?

Hello Alex.

I'm making a non-breaking improvement to a feature I've introduced before: https://github.com/argoproj/argo/pull/1350.
Previously, I've added a way to use per-workflow artifact location configmap. This enables different workflows or users to specify different artifact locations without having to hardcode all storage server details (endpoints, ports, secret keys) in the workflow.
Unfortunately, I made a mistake and used the workflow controller's namespace to locate the configmap. This make the following major user scenario impossible: each user has different namespace with their own secrets and artifact location configmaps. This PR fixes this deficiency. This does not break backwards compatibility since the lookup process falls back to the old behavior.

alexec commented:
Thank @Ark-kun - am I right in thinking your change was released in v2.4, but that you consider this a bug?

P.s. you should sync with master, as I don't think you have the config for running the Circle CI tests.

Ark-kun commented:
>Thank @Ark-kun - am I right in thinking your change was released in v2.4, but that you consider this a bug?

We've realized that the the feature failed to cover the biggest user scenario.
The feature released in 2.4.0 covered the scenario:
"Different users use same namespace, but differently named configMaps".
while a very big user scenario is actually
"Different users use different namespaces, with same-named configMaps".

>P.s. you should sync with master, as I don't think you have the config for running the Circle CI tests.

Thank you. Looks like it worked. This baffled me since usually the CI systems test the PR+master merge version.

Ark-kun commented:
@alexec The tests are now passing.

Ark-kun commented:
@alexec Gentle ping
